minima/database.nim

```diff
@@ -20,38 +20,55 @@ type
         KeyNotFound             = "minima: key not found"
         PersistenceFailed       = "minima: persistence failed"
 
-proc log(db: Database, key: seq[byte], value: seq[byte]) =
-    let write = concat(
-        @(uint32(len(key)).toBytes),
-        @(uint32(len(value)).toBytes),
-        key,
-        value
-    )
-
-    # @TODO CATCH EXCEPTIONS
-    discard db.log.writeBytes(write, 0, len(write))
-    db.log.flushFile()
+proc toAESKey*(str: string): array[aes256.sizeKey, byte] =
+    ## Concerts a string to an AES Key for opening an encrypted database.
+    ## 
+    ## **Example::**
+    ## .. code-block::
+    ##   let password = "foo".toAESKey
+    var pass = str
+    var key: array[32, byte]
+    copyMem(addr key[0], addr pass[0], len(pass))
 
-proc readInt(file: File): int =
-    var arr: array[4, byte]
-    discard file.readBytes(arr, 0, 4)
+    return key
 
-    return int(uint32.fromBytes(arr))
+proc open*(dir: string, key: array[aes256.sizeKey, byte]): Result[Database, DatabaseError] =
+    ## Opens an encrypted database at the specified path.
+    ## 
+    ## **Example:**
+    ##
+    ## .. code-block::
+    ##   let result = open("/tmp/minima", "password".toAESKey)
+    ##   if not result.isOk:
+    ##     return
+    if not os.existsDir(dir):
+        try:
+            os.createDir(dir)
+        except:
+            return err(DirectoryCreationFailed)
 
-proc recover(db: Database) =
-    while db.log.getFilePos() <= db.log.getFileSize() - 1:
-        var keyLen = readInt(db.log)
-        var valLen = readInt(db.log)
+    var path = dir & "/minima.db"
+    var mode = fmReadWrite
+    if fileExists(path):
+        mode = fmReadWriteExisting
 
-        # @TODO CATCH EXCEPTIONS
+    var f: File
```

Comment:
Often, you can make the code more elegant if you use try expressions:

```nim
var f = try: open(path, mode)
        except CatchableError: return err(TreeFileCreationFailed)
```

Furthermore, see how I'm catching `CatchableError`. With the current version of Nim, you should never write an unspecified `except: ` clause because it will handle `Defects` as well. A `Defect` is a non-recoverable error in the code (i.e. a bug, a logical error). It's the kind of error that can be fixed only by modifying the source code of the program. It's not expected to arise in a correctly implemented program and it should lead to an immediate crash, following the [Fail-fast principle](https://en.wikipedia.org/wiki/Fail-fast). 

On the other hand, a `CathableError` is something that is expected to arise even in a correctly implemented program and the developer is expected to make a decision specifying how the error is handled.

For a more complete explanation of Errors vs Defects, please see here:
https://gist.github.com/zah/d2d729b39d95a1dfedf8183ca35043b3#a-common-vocabulary-for-errors

Future version of Nim will treat `expect: ` as `except CatchableError: `, but that's not the case right now.